vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Collections.Generic
+open System.Collections.Immutable
+open System.Composition
+open Microsoft.CodeAnalysis.CodeFixes
+open Microsoft.CodeAnalysis.Text
+open Microsoft.VisualStudio.FSharp.Editor.Extensions
+open CancellableTasks
+
+[<AutoOpen>]
+module private Patterns =
+    let inline toPat f x = if f x then ValueSome() else ValueNone
+
+    [<return: Struct>]
+    let inline (|LetterOrDigit|_|) c = toPat Char.IsLetterOrDigit c
+
+    [<return: Struct>]
+    let inline (|Punctuation|_|) c = toPat Char.IsPunctuation c
+
+    [<return: Struct>]
+    let inline (|Symbol|_|) c = toPat Char.IsSymbol c
+
+module private SourceText =
+    /// Returns true if the given span contains an expression
+    /// whose indentation would be made invalid if the open paren
+    /// were removed (because the offside line would be shifted), e.g.,
+    ///
+    ///     // Valid.
+    ///     (let x = 2
+    ///      x)
+    ///
+    ///     // Invalid.
+    ///    ←let x = 2
+    ///      x◌
+    ///
+    ///     // Valid.
+    ///     ◌let x = 2
+    ///      x◌
+    let containsSensitiveIndentation (span: TextSpan) (sourceText: SourceText) =
+        let startLinePosition = sourceText.Lines.GetLinePosition span.Start
+        let endLinePosition = sourceText.Lines.GetLinePosition span.End
+        let startLine = startLinePosition.Line
+        let startCol = startLinePosition.Character
+        let endLine = endLinePosition.Line
+
+        if startLine = endLine then
```

Comment:
Hmm, there are indeed probably other scenarios where it'd be useful to be aware of potential offside-line skewing, but this code is currently at least partly parentheses-specific, and for other uses you'd probably need to make it agnostic of open/close tokens, open/close token length, etc.: for example, maybe you couldn't just use `IndexOfAnyExcept(' ', '(')`, etc., since the open or close token is a full word or whatever, and so on. I think I'll leave that for another day :)

Reply:
Alrigjhty, let's keep it here and split it if/once necessary.